### PR TITLE
fix(examples/fsdp_tp): make --compile work; exact MFU; drop no-op flags

### PR DESCRIPTION
**Pull request by saforem2**
Follow-up to #166 — a set of `examples/fsdp_tp.py` (+ `flops.py`, `vit.py`, `diffusion.py`) fixes that were developed against that branch but never committed before it was squash-merged. Rebased cleanly onto current `main`.

### 1. `feat(flops)` — exact fake-tensor MFU estimator + lower-bound fix
The old MFU path measured FLOPs at seq=128 and linearly scaled to the real seq. On CPU and on the fused SDPA backends (flash/efficient/cuDNN), `FlopCounterMode` reports **zero** for the attention op, so both the probe and the real shape silently dropped the entire O(seq²) attention term. Reported MFU came out ~36% low for agpt-xl @ seq=8192 (verified: 1.19e13 vs 1.85e13 FLOPs).

New `estimate_model_flops_fake()` runs fwd+bwd under `FakeTensorMode` (shape-only, no OOM at real seq) with `sdpa_kernel(MATH)` forced so SDPA decomposes into countable `bmm`s. `fsdp_tp.py` prefers it, falling back to the linear probe. The fallback log message is corrected: under-counting makes MFU a **lower** bound, not an upper bound.

### 2. `refactor(examples)` — delete 5 no-op flags + backfill help text
Removed flags argparse parsed but no code read: `--model_name_or_path`, `--depth-init` (fsdp_tp); `--fsdp-mixed-precision` (diffusion, superseded by `--dtype`); `--hidden-dim`, `--mlp-dim` (vit, derived from `num_heads*head_dim`). Passing them now errors instead of silently no-op'ing. Also added `help=` to the ~20 bare fsdp_tp flags.

### 3. `fix(examples/fsdp_tp)` — make `--compile` work with TP + FSDP + AC
Three independent `torch.compile` blockers fixed:
- **Per-block compile** (torchtitan's `apply_compile` pattern) instead of whole-model — avoids the Dynamo `_MaskPartial` graph break on TP-wrapped `tok_embeddings`.
- **`ptd_checkpoint_wrapper`** for activation checkpointing instead of a monkey-patched `torch.utils.checkpoint.checkpoint` closure — the closure graph-breaks under `fullgraph=True` (gb0053).
- **`use_orig_params=True`** on both FSDP1 wrap sites — Dynamo refuses `use_orig_params=False`.

⚠️ **Known upstream limitation:** `{--tp>1, --activation-checkpoint, --compile}` *all three at once* still trips an AOTAutograd assertion (a `DeviceMesh` leaks into `tensors_saved_with_vc_check`). Not fixable here; repro lives in torchtitan `docs/upstream-issues/repro_devicemesh_in_saved_tensors.py`. Workaround: drop any one. Every other compile combination now works.

### 4. `feat(examples/fsdp_tp)` — surface `--sharding-strategy` choices + help
It was the one closed-set flag rendering bare in `--help`. Added `choices=` (so argparse validates + prints the set) and a help string mapping each value to its ZeRO stage.

## Verification
- All 4 modules import; `--help` renders; 5 deleted flags absent.
- Compile-flag CLI contract (the `test_examples_compile_flag.py` cases) passes for all 5 examples.
- Numeric flops check: fake-tensor exact = 1.85e13 vs linear-scaled 1.19e13 (35.7% attention recovered); MFU(exact) > MFU(under-count) confirms lower-bound semantics.
- End-to-end: AC(full) + per-block `fullgraph=True` compile + fwd/bwd runs clean.

> Note: pytest couldn't run on this Sunspot UAN host (repo `.venv` targets an Aurora-only spack python; the working torchtitan venv lacks pytest/pip). Contracts were verified with standalone stdlib harnesses replicating the suite assertions. A CI run on the PR will exercise the full pytest suite.

## Summary by Sourcery

Improve FLOPs estimation and torch.compile compatibility for the FSDP+TP example while cleaning up unused CLI flags across examples.

New Features:
- Add a fake-tensor-based FLOPs estimator that counts attention exactly at real batch/sequence shapes and expose it to examples.
- Expose FSDP sharding strategies as validated CLI choices with descriptive help text in the FSDP+TP example.

Bug Fixes:
- Make torch.compile work reliably with tensor parallelism, FSDP, and activation checkpointing by switching to per-block compilation, using FSDP use_orig_params, and adopting the ptd checkpoint wrapper instead of a checkpoint closure.
- Remove no-op/unused CLI flags from the FSDP+TP, ViT, and diffusion examples so invalid arguments now surface as errors rather than silently doing nothing.

Enhancements:
- Document and clarify most FSDP+TP example CLI arguments with detailed help strings for usability.
- Refine FLOPs logging and MFU semantics so linear-scaling estimates are treated and messaged as lower bounds when attention is under-counted.

### CODE REVIEW

**Comment by saforem2** [comment]
## ✅ Test suite run on Sunspot compute node

Ran the suite on a dedicated 1-node PBS allocation (`ezpz_setup_env` → python 3.12 + torch 2.10 + pytest 9.0.2), since the repo `.venv` targets an Aurora-only spack python not mounted on the UAN.

### Targeted suites (the code this PR changes) — all green
```
tests/test_examples_compile_flag.py
tests/test_fsdp_tp_activation_checkpoint.py
tests/test_examples_preset_overrides.py
tests/test_flops.py
→ 117 passed, 4 warnings in 35.00s
```

### Full suite — no regressions from this PR
```
1170 passed, 9 skipped, 2 failed   (-p no:randomly)
```

The **2 failures are pre-existing and unrelated to this PR**:
- `test_yeet_env.py::TestRun::test_generic_source_skips_venv_footer`
- `test_yeet_env.py::TestRun::test_venv_source_uses_venv_footer`

Both assert on captured **stdout** (`capsys`) for the yeet footer, but that output now routes through `ezpz.get_logger` (since #161's `fc76fd1`), so `capsys.readouterr().out` is empty. `src/ezpz/utils/yeet_env.py` and `tests/test_yeet_env.py` are **byte-identical between `origin/main` and this branch** (this PR touches only `flops.py` + 3 example files), so the failures reproduce on `main` independently. Worth a separate fix, out of scope here.

### Note on the 2116 "errors" if you run without `-p no:randomly`
The `pytest-randomly` reseed hook calls `deepspeed.set_random_seed` → `torch.manual_seed` → `torch.xpu._is_in_bad_fork()`, and `torch.xpu` raises `AttributeError` on this Aurora torch-2.10 build — firing at setup/teardown of every deepspeed-importing test. These are environment/plugin errors (0 test-logic failures), not regressions. `-p no:randomly` avoids them.

File: src/ezpz/examples/fsdp_tp.py [hunk 1]
```diff
@@ -710,7 +893,6 @@ def parse_args(argv: Optional[list[str]] = None):
     )
     # max_batch_size: int = 32
     # max_seq_len: int = 32768
-    # depth_init: bool = True
     args = parser.parse_args(argv)
     apply_model_preset(args, argv)
     return args
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:898]: Correct — the old `torch.utils.checkpoint.checkpoint` path defaulted to `preserve_rng_state=True`, so passing `False` here changed dropout/RNG semantics on recompute. Changed both call sites to `preserve_rng_state=True` in 51cee38 to restore the prior (correct) behavior.

File: src/ezpz/flops.py [hunk 2]
```diff
@@ -321,6 +323,129 @@ def try_estimate(
         return 0
 
 
+def estimate_model_flops_fake(
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/flops.py:326]: Partially applied: narrowed the try-region and moved the imports inside it in 51cee38 (also addresses the Copilot import-safety note on the same function). Deferring the broader `_build_dummy_input` / `_count_flops_with_counter` / `_try_estimate_common` extraction — it's a reasonable cleanup but the duplication is ~15 lines across two functions and the shared-wrapper indirection trades a bit of readability for it. Leaving as a follow-up rather than expanding this PR's surface.

File: src/ezpz/examples/fsdp_tp.py [hunk 3]
```diff
@@ -1146,38 +1334,57 @@ def train(
     logger.info(f"\n{mstr}")
     model.to(device)
 
-    # Run try_estimate with a tiny probe (batch=1, seq=128) and scale up
-    # by total-tokens ratio. The full (batch_size, seq_len) shape would
-    # OOM on big models — FlopCounterMode does a real forward pass, and
-    # a 2B-param model at seq=8192 materializes multi-GiB of activations
-    # per rank before FSDP has a chance to shard.
+    # FLOPs estimation: try the exact fake-tensor path first, fall back
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:1343]: Acknowledged. The `estimate_flops_with_fallback` and `apply_blockwise_compile` extractions are sensible and would slim `train()`. Deferring to a focused follow-up — this PR is already a multi-concern change and I'd rather not reshuffle `train()`'s control flow in the same diff. Noted for a cleanup pass. (One caveat on the suggested `apply_blockwise_compile`: reusing `_find_block_list` would also pick up the HF deepest-ModuleList fallback, which the compile path intentionally doesn't want — it keys on `model.layers` / `model.model.layers` specifically. Would need to preserve that distinction.)

File: src/ezpz/flops.py [hunk 4]
```diff
@@ -321,6 +323,129 @@ def try_estimate(
...
+        input_shape: Real shape of the input tensor (e.g.
+            ``(batch, seq_len)``). No memory allocated at this shape.
+        backward: If ``True``, also count the backward pass FLOPS.
+
+    Returns:
+        Total FLOPS (forward + backward if requested), or 0 on failure.
+    """
+    from torch._subclasses.fake_tensor import FakeTensorMode
+    from torch.nn.attention import sdpa_kernel, SDPBackend
+
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/flops.py]: Fixed in 51cee38. Moved the `FakeTensorMode` / `sdpa_kernel` imports inside the try block so a build lacking them returns 0 per the docstring contract instead of raising ImportError. Added a test (`test_returns_zero_when_fake_mode_unavailable`) that monkeypatches the import to fail and asserts 0 is returned.

File: src/ezpz/flops.py [hunk 5]
```diff
@@ -321,6 +323,129 @@ def try_estimate(
...
 
 
+def estimate_model_flops_fake(
+    model: torch.nn.Module,
+    input_shape: tuple[int, ...] | list[int],
+    *,
+    backward: bool = True,
+) -> int:
+    """Count FLOPS exactly via FakeTensorMode — no memory, no OOM.
+
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/flops.py:333]: Added in 51cee38. `tests/test_flops.py` now has `TestEstimateModelFlopsFake` + `TestTryEstimateFake`: returns-positive-int, embedding vs float input paths, backward>forward, the wrapper catches exceptions (returns 0), and returns 0 when FakeTensorMode is unavailable. Full file: 43 passed.

File: src/ezpz/examples/fsdp_tp.py [hunk 6]
```diff
@@ -893,38 +1095,24 @@ def _find_block_list(m: nn.Module) -> Optional[nn.ModuleList]:
...
-            block.forward = checkpointed_forward  # type: ignore[assignment]
-            return block
+    if mode == "block":
+        # Replace each block with a CheckpointWrapper(block) in-place on
+        # the ModuleList. Subsequent per-block torch.compile sees the
+        # wrapped modules and can trace through them.
+        for i in range(len(blocks)):
+            blocks[i] = ptd_checkpoint_wrapper(
+                blocks[i], preserve_rng_state=False
+            )
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:1111]: Tested this directly on hardware (Sunspot, 6× PVC, torch 2.14+xpu) and it does **not** reproduce as described. `CheckpointWrapper` proxies attribute access through `__getattr__`, so `block.attention` / `block.feed_forward` (and thus `_register_activation_hooks`) still resolve to the real submodules after wrapping — no AttributeError, and the run does not raise before training.

Ran `--tp 2 --ac block` with `EZPZ_TRACK_HIST=1 EZPZ_TRACK_ACT_HIST=1` (no compile): it trained cleanly past 3000 iterations with the histogram hooks active, loss 10.9→4.9. So AC-block + activation-hist coexist fine.

There *is* a real but different incompatibility: the histogram hooks break under `--compile` (not under AC) — the hook's `activations[tag] = out.detach()` dict mutation is an unsupported side-effect for Dynamo under `fullgraph=True` (gb0067). That's a hooks-vs-compile issue, independent of the CheckpointWrapper change, and predates this PR's AC rewrite. Not addressing it here; the AC change itself is sound.
